### PR TITLE
fingering layout

### PR DESCRIPTION
**Pull request by MarcSabatella**
There are quite a few issues involving fingering layout.  Some are design limitations, some are bugs inherited from 2.3.2, some are some regressions.  This PR tries to address them all as well as possible.

Some of the issues fixed:

- All sources agree that piano fingerings for both single notes and chords should be stacked above the chord, but we have always placed them to the left of noteheads because it was too hard to stack them correctly.  Fingerings for single notes we place above the chord but not necessarily above the staff as we should.  Pianists have relied on a plugin to fix the fingering, but that's not an option right now.
- 2.3.2 did a good job of placing piano fingerings above beamed groups of notes, but 3.0 gets this wrong, overlapping the stem in many cases.
- In 2.3.2, we placed LH guitar fingering to the left of notes as per the usual standard, but 3.0 only does this for chords, not single notes (we use the same algorithm for LH guitar fingering as for piano).  Guitar players have rightly complained about this.
- User styles would be useful to set custom offsets from notes, but that works only if autoplace leaves these in the default position.

No doubt there is still room for improvement even with my fixes, but I think mostly it will be about small tweaks - quibbling over exactly how far from the notehead the fingering is, etc.  I'd like to see this merged and let people like cadiz1 and geetar have at it to provide more feedback.

One limitation I know we will probably want to revisit is the vertical alignment and spacing of LH guitar fingering in chords.  With the possibility of tightly-spaced chords (including seconds), plus accidentals, it's going to be next to impossible to come up with perfect layout by default in all cases.  My algorithm here does pretty well for many chords but will still have collisions in some cases.  That's probably unavoidable, hopefully people find my approach avoids the need for manual adjustments in many cases and provides a good starting point in the rest.

Here's a sample:

image

### CODE REVIEW

**Comment by MarcSabatella** [comment]
The broken tests are kind of to be expected as a side effect of how I implemented the algorithm, making use of the placement property.  I could update the references, but frankly, I'd rather update the writing of Fingering elements to skip writing this property, because it is not actually needed (it is generated automatically).  Hmm, I was thinking that would mean creating an override for Fingering::writeProperties, but actually, maybe I can avoid that by forcing this property to always be "styled" (it is not exposed in the UI).  Anyhow, one way or another I will fix.

**Comment by MarcSabatella** [comment]
FWIW, I updated the references.  That way if we ever *do* decide to expose the placement property for fingering, the tests won't break again.

**Comment by MarcSabatella** [comment]
Updated to incorporate some excellent feedback from cadiz1 and geetar in the issue report https://musescore.org/en/node/280807

**Comment by MarcSabatella** [comment]
In anticipation of needing to close issues manually once this is merged, here's a list of issues addressed here:

https://musescore.org/en/node/280807 (as per commit message)
https://musescore.org/en/node/280088
https://musescore.org/en/node/279113
https://musescore.org/en/node/280412
https://musescore.org/en/node/280516
https://musescore.org/en/node/280078
https://musescore.org/en/node/280017
https://musescore.org/en/node/279241
https://musescore.org/en/node/274936
https://musescore.org/en/node/267499
https://musescore.org/en/node/115861
https://musescore.org/en/node/279963

File: libmscore/chord.cpp [hunk 1]
```diff
@@ -1985,6 +1995,30 @@ void Chord::layoutPitched()
...
+                                    leftFound = true;
+                                    qreal xf = f->ipos().x();
+                                    xNote = qMin(xNote, xf);
+                                    }
+                              }
+                        }
+                  }
+            }
+      for (Fingering* f : alignNote) {
+            if (f->autoplace())
```
Comment (JoshuaBonn1) [inline libmscore/chord.cpp]: Is this needed if you already checked the elements earlier for autoplace? Or does autoplace have a side effect?

Comment (MarcSabatella) [reply libmscore/chord.cpp]: Good catch!   Harmless, but I can remove this...

**Comment by MarcSabatella** [comment]
Build failure doesn't seem to be my fault, AppVeyor may be full?

File: libmscore/chord.cpp (hunk 1 above)
Comment (MarcSabatella) [reply libmscore/chord.cpp]: (fixed)

**Comment by Jojo-Schmitz** [comment]
Force a new build and it should work

**Pull request merged**